go/pkg/cas/client.go

```diff
@@ -0,0 +1,96 @@
+// Package cas implements an efficient client for Content Addressable Storage.
+package cas
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+	"google.golang.org/grpc"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/client"
+	repb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+)
+
+// Client is a client for Content Addressable Storage.
+// Create one using NewClient.
+//
+// Goroutine-safe.
+//
+// All fields are considered immutable, and should not be changed.
+type Client struct {
+	// The full name of the RBE instance.
+	InstanceName string
+
+	// The configuration that the client was created with.
+	ClientConfig
+
+	conn       *grpc.ClientConn
+	byteStream bspb.ByteStreamClient
+	cas        repb.ContentAddressableStorageClient
+}
+
+// ClientConfig is a config for Client.
+// See DefaultClientConfig() for the default values.
+type ClientConfig struct {
+	DialParams client.DialParams
+
+	// TODO(nodir): add conifg values here.
+	// TODO(nodir): add per-RPC timeouts.
+	// TODO(nodir): add retries.
+}
+
+// DefaultClientConfig returns the default config.
+//
+// To override a specific value:
+//   cfg := DefaultClientConfig()
+//   ... mutate cfg ...
+//   client, err := NewClientWithConfig(ctx, cfg)
+func DefaultClientConfig() ClientConfig {
+	return ClientConfig{
+		DialParams: client.DialParams{
+			Service: "remotebuildexecution.googleapis.com:443",
+		},
+	}
+}
+
+// Validate returns a non-nil error if the config is invalid.
+func (c *ClientConfig) Validate() error {
+	if c.DialParams.Service == "" {
+		return errors.Errorf("DialParams.Service is unspecified")
+	}
+
+	return nil
+}
+
+// NewClient creates a new client with the default configuration.
+func NewClient(ctx context.Context, instanceName string) (*Client, error) {
+	return NewClientWithConfig(ctx, instanceName, DefaultClientConfig())
+}
+
+// NewClientWithConfig creates a new client and accepts a configuration.
+func NewClientWithConfig(ctx context.Context, instanceName string, config ClientConfig) (*Client, error) {
```

Comment:
Should we also have a constructor that takes a `grpc.ClientConn`?

I imagine the long-term vision for this package is for `go/pkg/client` to use it as a dependency. In which case I think it may be better to just downstream that connection rather than create a new one.

Reply:
Why not accept the connection via Config, since it is an optional param?

Reply:
That also sounds good to me (and maybe add some comments and/or error checking that validates that setting ClientConfig.Conn is incompatible with setting `DialParams`).

Reply:
Added Conn. Clarified in the comments that Conn overrides DialParams.

Making the two fields mutually exclusive is a bit awkward:
1. Callsite with an existing connection would have to write

  ```go
    cfg := cas.DefaultClientConfig()
    cfg.DialParams = nil  // the awkward part
    cfg.Conn = myConn
    client, err := cas.NewClientWithConfig(ctx, instance, cfg)
  ```

2. It requires `ClientConfig.DialParams` to be a pointer, but it is not a big deal. WDYT?

Reply:
I would prefer the current version where Conn overrides DialParams. Maybe log a warning message if both are set?

Reply:
Ramy, just to make sure we are on the same page, do you prefer the callsite having to do both 1) set `Conn` and 2) unset `DialParams`; in order to avoid a warning? See my message right above for more detials.

Reply:
Ah sorry I missed the options above. Yes, 1 is awkward, perhaps we can have a DefaultWithDialParams or DefaultDialParams?

Reply:
Could you elaborate? Do you mean three symbols

- `DefaultClientConfigWithParams` which returns a fully-functioning config with `DialParams` and without `Conn`
- `DefaultDialParams` used by above
- `DefaultClientConfig` which returns a non-really-functioning client config because it has neither `DialParams` nor `Conn`? It is primarily for those who want to set `Conn`.

This seems a bit awkward to me as well, in particular the fact that a single attribute in the config causes symbol explosion at the package level.

Maybe it is worth to step back a bit to think about the problem we want to solve. FWIU it is that a programmer might get surprised that `DialParams` is ignored. I wonder if this can be addressed with a louder comment? Sort of "warning: if you override Conn, then DialParams is ignored" or something.

Reply:
I meant either  `DefaultDialParams` + `DefaultClientConfig` or `DefaultClientConfigWithParams` + `DefaultClientConfig`. In the first case one would:
```
cfg := cas.DefaultClientConfig()
cfg.DialPararms = cas.DefaultDialParams()
```

Yes that is the concern. The user could override some `DialParams` for CAS versus Exec, but accidentally sharing the connection would silently override the changes. I guess since overriding the connection will be explicit, the user should be aware of what they're doing. So a comment sounds sufficient to me.

Reply:
> I guess since overriding the connection will be explicit, the user should be aware of what they're doing

yes, I implied this. Sorry for not making this explicit.

I took a stab at a louder version of the comment. PTAL.